### PR TITLE
Perl 6 -> Raku rename

### PR DESCRIPTION
**Pull request by coke**
A first pass at converting Perl6 to Raku.

### CODE REVIEW

**Comment by coke** [comment]
I couldn't see how to add this to the 'make units' test; everything seems to get built and all other tests pass, however.  Please review and let me know if I can do anything to improve the PR.

**Comment by coke** [comment]
Once this is approved, someone (me?) can work on updating any syntax changes since the original Perl 6 version, and perhaps removing the code to sniff out the differences between 5 & 6 (which are now using different extensions entirely)

File: win32/ctags_vs2013.vcxproj [hunk 1]
```diff
@@ -477,4 +477,4 @@
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
   <ImportGroup Label="ExtensionTargets">
   </ImportGroup>
-</Project>
```
Comment (masatake) [inline win32/ctags_vs2013.vcxproj:480]: Don't add the newline at the end of file.

File: parsers/raku.c [hunk 2]
```diff
@@ -321,21 +322,20 @@ findPerl6Tags (void)
...
+    deinitRakuCtx(&ctx);
 }
 
 parserDefinition *
-Perl6Parser (void)
+RakuParser (void)
 {
-    static const char *const extensions[] = { "p6", "pm6", "pm", "pl6",
-		"t6", "raku", "rakumod", "rakutest", NULL };
+    static const char *const extensions[] = { "raku", "rakumod", "rakutest", "rakudoc", NULL };
```
Comment (masatake) [inline parsers/raku.c:331]: ctags should support older code.
With this change, ctags cannot handle Perl6 files having '.pl6' as a suffix.
This is degrading.

**Comment by masatake** [comment]
Can we assume Raku is a superset of Perl6?

This change may break two things:

1. ctags cannot read *.pl6 files.
2. options of ctags command lines like --kind-Perl6=+c don't work.

I found two approaches:

A. add a new file, raku.c, by copying perl6.c.
   - then revise raku.c to the latest Raku syntax, and
   - remove Raku-specific code from perl6. e.g. `static const char *const extensions[] = { ... , "raku", "rakumod", "rakutest", NULL };`

B. add a `parserDefinition` for Raku to perl6.c like:
```
diff --git a/parsers/perl6.c b/parsers/perl6.c
index e455642e5..d691e6828 100644
--- a/parsers/perl6.c
+++ b/parsers/perl6.c
@@ -327,15 +327,31 @@ findPerl6Tags (void)
 parserDefinition *
 Perl6Parser (void)
 {
-    static const char *const extensions[] = { "p6", "pm6", "pm", "pl6",
-		"t6", "raku", "rakumod", "rakutest", NULL };
+    static const char *const extensions[] = { "p6", "pm6", "pm", "pl6",NULL };
     static selectLanguage selectors [] = { selectByPickingPerlVersion,
 					   NULL };
     parserDefinition* def = parserNew("Perl6");
     def->kindTable      = perl6Kinds;
     def->kindCount  = ARRAY_SIZE(perl6Kinds);
     def->extensions = extensions;
+	def->aliases = aliases;
     def->parser     = findPerl6Tags;
     def->selectLanguage = selectors;
     return def;
 }
+
+parserDefinition *
+RakuParser (void)
+{
+    static const char *const extensions[] = { "raku", "rakumod", "rakutest", NULL };
+    static selectLanguage selectors [] = { selectByPicking?Version,
+					   NULL };
+    parserDefinition* def = parserNew("Raku");
+    def->kindTable      = ?Kinds;
+    def->kindCount  = ARRAY_SIZE(?Kinds);
+    def->extensions = extensions;
+	def->aliases = aliases;
+    def->parser     = find?Tags;
+    def->selectLanguage = selectors;
+    return def;
+}
```

`?` are debatable.

If we should think perl6 and raku are different languages, we should choose A.
If not, B can be a choice.

@dtikhonov Do you have comments?

**Comment by coke** [comment]
Perl6 no longer exists, since 2019. It was renamed to Raku. So, not two different languages.

Despite that, I can keep the old extensions.

I’ll undo the ws change, sorry about that.

**Comment by masatake** [comment]
I guess Raku doesn't accept '*.pm' So you don't have to change selectors.c
selectors.c is for arbitrating between Perl6 and Perl; both two can handle '.pm' files.

Your change looks o.k.
I would like to adjust RakuParser and keep Perl6Parser like:

```
parserDefinition *
RakuParser (void)
{
    static const char *const extensions[] = { "raku", "rakumod", "rakutest", NULL };
    parserDefinition* def = parserNew("Raku");
    def->kindTable      = raku6Kinds;
    def->kindCount  = ARRAY_SIZE(rakuKinds);
    def->extensions = extensions;
    def->parser     = findRakuTags;
    return def;
}

/* For keeping the name "Perl6". */
#define perl6Kinds rakuKinds /* TODO: kinds should not be shared the kind spec between parsers. */
#define findPerl6Tags raku6Kinds
parserDefinition *
Perl6Parser (void)
{
    static const char *const extensions[] = { "p6", "pm6", "pm", "pl6", "t6", NULL };
    static selectLanguage selectors [] = { selectByPickingPerlVersion,
					   NULL };
    parserDefinition* def = parserNew("Perl6");
    def->kindTable      = perl6Kinds;
    def->kindCount  = ARRAY_SIZE(perl6Kinds);
    def->extensions = extensions;
    def->parser     = findPerl6Tags;
    def->selectLanguage = selectors;
    return def;
}
```

**Comment by masatake** [comment]
BTW, you use your `master` branch for making a pull request.
I think you should make a topic branch.

**Comment by dtikhonov** [comment]
No, I don't have comments about the code change, @masatake.

**Comment by masatake** [comment]
I pushed commits reflecting my review comments to your branch.
I think this is ready for merging after squashing the commits into one.

@coke Will you modify or update this pull request more before merging?

**Comment by coke** [comment]
I am **very** surprised that you would push into my fork on the branch I was working on. I didn't even realize this would be allowed by default.

Please merge what you like: if you're happy with the current state. I do note that "make units | grep -i raku" still shows no tests being run.

**Comment by coke** [comment]
> I am **very** surprised that you would push into my fork on the branch I was working on. I didn't even realize this would be allowed by default.

Looking at settings on coke/ctags, no one else is permitted; I'm very confused.

**Comment by coke** [comment]
> I'm very confused.

When opening a PR, "allow edits by maintainers" setting is apparently enabled by default.

**Comment by masatake** [comment]
>  I do note that "make units | grep -i raku" still shows no tests being run.

Thank you. I didn't recognize it.
I inspected it.

I guess you did `git mv Units/parser-perl6.r Unit/parser-raku`.
This should be `git mv Units/parser-perl6.r Unit/parser-raku.r`.

**Comment by masatake** [comment]
image

**Comment by masatake** [comment]
Thank you. I'm waiting for more contributions from you:-).
I hope https://docs.ctags.io/en/latest/testing-parser.html may help you improve the parser.

> I am very surprised that you would push into my fork on the branch I was working on. I didn't even realize this would be allowed by default.

I was very surprised by the feature, too.

I have used the feature only when I want to merge a pull request quickly.
I should not use "push --force" with this feature not to break contributors' work critically.
"git push" can keep all the original commits on the branch. Instead, I use the "Squash and Merge" button on GitHub.

**Comment by coke** [comment]
> This should be `git mv Units/parser-perl6.r Unit/parser-raku.r`.

Good catch!

**Comment by coke** [comment]
With the tests passing, let me know if there's anything else needed for this PR. Thanks!

**Pull request merged**